### DISCUSSION TITLE
Badges Overhaul

### DISCUSSION THREAD

**Discussion by zack-o-zack** [typehero/typehero#1304] [Ideas]
# Badges Overhaul

In the current setting Badges looks like a non-important part of the platform. Users don't know about how many badges they have earned (until they open their Profile), also needs other improvements overall.

Suggestions:
1. Asynchronous Loading of badges
_Currently, we load badges like this. This may lead to longer wait times if we add more badges._
```ts
const badges = await getBadges(user.id);

  return (
    <Dashboard user={user} isOwnProfile={isOwnProfile} badges={badges}>
      {children}
    </Dashboard>
  );
```
2. Add Badges Skeleton

3. Show badges on **Successful Submission** (AOT 2023 Related)
_This will directly show the user that they have earned a Badge without any schema changes / adding notification feature.(Only AOT / other events)_
```ts
<Badges user={user} category={['aot-2023']} />
```

4. New Badges
_Ideas are appreciated, I only thought about adding a `new` **'Contributors'** Badge_ 
![image](https://github.com/typehero/typehero/assets/69889382/5bafa47e-b332-4f9a-abaa-1db9229af0a5)
Figma File: [TypeHero Badges](https://www.figma.com/file/yXLUFMSfluUHsBsBHT4t3d/TypeHero-Badges?type=design&node-id=0%3A1&mode=design&t=qhixt85pJn0FjjPI-1)

5. Badges Lore
_Clicking on badges opens a dialog, that has additional details, and an option to share their achievement on Twitter_
![image](https://github.com/typehero/typehero/assets/69889382/7ef5c5bb-2b5c-462c-8284-78d54b2140c0)
 
6. New Badge Schema (needs improvement)
```ts
export interface BadgeInfo {
  // Badge category
  category: 'aot-2023' | 'contributor';
  // Unique Identifier for a badge
  slug: string;
  // Badge name to be visible in UI
  name: string;
  // Badge level (Priority over other badges having the same category, used while grouping)
  level?: number;
  //  (Tier of badge, visible on UI)
  // eslint-disable-next-line @typescript-eslint/sort-type-constituents
  tier?: 'bronze' | 'silver' | 'gold' | 'platinum';
  // Eligibility to earn the badge, Visible on UI.
  eligibility: string;
```

7. Grouping Badges (AOT 2023 Related)
_Group similar badges on UI, e.g. All AOT 2023 Badges together._
![image](https://github.com/typehero/typehero/assets/69889382/9a06af43-c8b6-42fe-ac01-02ac0cfd783f)

**Comment by bautistaaa** [comment]
ohhh i like it. the github badge.

i think contributors/admin/mod should also have "flair" next to their user name so they stand out in the comments as well

**Reply by zack-o-zack** [reply]
If we add contributor to the roles. It'll be easy after that.

**Comment by CodeItQuick** [comment]
What are some examples of potential badges? To brainstorm some good/bad ideas, here's a starter list:
1. Moderator Badge
2. Contributor Badge
3. Typescript Foundation Badge
4. Badge for each challenge completed (?)
5. Advent of Typescript 2023/Advent of Typescript 2024 (maybe break these down into completed 5 days, 10 days, etc.)

To expand on Advent of Typescript, maybe something like
(Bronze) 5 days completed
(Bronze) 10 days completed
(Silver) 15 days completed
(Silver) 20 days completed
(Gold) 25 days completed

For what its worth this is dota2attitude, was thumbing around in the issues looking for stuff I might want to pickup.

**Comment by CodeItQuick** [comment]
I'm going to grab this one.
